### PR TITLE
JIT: profile updates for finally optimizations

### PR DESCRIPTION
**Pull request by AndyAyersMS**
Update `impImportLeave` to propagate IBC counts to new blocks.

Fix profile weights during callfinally chain merging.

Scale profile weights for cloned finallys. Choose the continuation path based
on profile weight. Make sure to keep handler entry hot.

Partial fix for #48925.

### CODE REVIEW

**Comment by AndyAyersMS** [comment]
cc @dotnet/jit-contrib 

No diffs for non-pgo. 

~ 90 methods impacted in SPMI collection

Sample diff. In this method there are 3 exits from the try-finally. The two early ones (BB04 and BB06) have the same continuation; the lexically last exit BB08 has a different continuation. Before the changes in this PR we would reroute flow from BB08 to the clone, but profiling shows that we should prefer the other exit path.

Also note in the before picture the step blocks (BB12-BB16) have no profile counts, and both the original finally BB09 and the cloned finally BB18 have the same count, and the cloned finally is reached via BB08 which has zero profile count.

In the after picture the step blocks have IBC counts (from the change to `impImportLeave`), the cloned finally is reached via BB04/BB06, and the original finally has no IBC count (it should have count zero, but because of the "handler entry must be hot" logic it is given a nonzero non-IBC weight).

<table>
<tr><td>Before</td><td>After</td>
<tr> <td valign=top>
<image src="https://user-images.githubusercontent.com/10121823/109998343-fe31a500-7cc5-11eb-8a37-4e77ac8df7eb.png" width = 380" />
</td><td valign=top>
<image src="https://user-images.githubusercontent.com/10121823/109998221-dcd0b900-7cc5-11eb-9cd1-f9643de342e7.png" width = 380" />
</td></tr></table>

**Comment by AndyAyersMS** [comment]
Net diff is slightly positive...
```
Total bytes of delta: -519 (-0.57% of base)
    diff is an improvement.
...
91 total files with Code Size differences (59 improved, 32 regressed), 31 unchanged.
```

File: src/coreclr/jit/fgehopt.cpp [hunk 1]
```diff
@@ -1067,9 +1135,10 @@ PhaseStatus Compiler::fgCloneFinally()
         // Modify the targeting call finallys to branch to the cloned
         // finally. Make a note if we see some calls that can't be
         // retargeted (since they want to return to other places).
-        BasicBlock* const firstCloneBlock    = blockMap[firstBlock];
-        bool              retargetedAllCalls = true;
-        BasicBlock*       currentBlock       = firstCallFinallyRangeBlock;
+        BasicBlock* const    firstCloneBlock    = blockMap[firstBlock];
+        bool                 retargetedAllCalls = true;
+        BasicBlock*          currentBlock       = firstCallFinallyRangeBlock;
+        BasicBlock::weight_t retargetedWeight   = 0;
```
Comment (BruceForstall) [inline src/coreclr/jit/fgehopt.cpp:1141]: 0 =>  BB_ZERO_WEIGHT

**Review by BruceForstall — APPROVED** [review]
LGTM

File: src/coreclr/jit/fgehopt.cpp (hunk 1 above)
Comment (AndyAyersMS) [reply src/coreclr/jit/fgehopt.cpp]: Yup.

**Comment by AndyAyersMS** [comment]
Another occurrence of #11063
```
Unhandled exception. System.InvalidOperationException: Collection was modified after the enumerator was instantiated.
```

**Pull request merged**